scripts/docs_link_checker.py

```diff
@@ -0,0 +1,306 @@
+        relative_link_regex = r"^(?P<path>\.\.?[\.\w\/-]+\.md)(?:#\S+)?$"
+        self._relative_link_pattern = re.compile(relative_link_regex)
+
+        absolute_image_regex = r"^\/" + site_prefix + r"\/(?P<path>[\w\/-]+\.\w{3,4})$"
+        self._absolute_image_pattern = re.compile(absolute_image_regex)
+
+        # ending with a 3-4 character suffix
+        relative_image_regex = r"^(?P<path>\.\.?[\.\w\/-]+\.\w{3,4})$"
+        self._relative_image_pattern = re.compile(relative_image_regex)
+
+    def _is_image_link(self, markdown_link: str):
+        return markdown_link.startswith("!")
+
+    def _is_doc_link(self, markdown_link: str):
+        return not self._is_image_link(markdown_link)
+
+    def _is_anchor_link(self, link: str):
+        return link.startswith("#")
+
+    def _check_external_link(self, link: str, file: str) -> LinkReport:
+        if self._skip_external:
+            return None
+
+        logger.debug("Checking external link %s in file %s", link, file)
+
+        try:
+            response = requests.get(link)
+
+            if 400 <= response.status_code < 500:
+                logger.info(
+                    "External link %s failed in file %s with code %i",
+                    link,
+                    file,
+                    response.status_code,
+                )
+                return LinkReport(
+                    link,
+                    file,
+                    f"External link returned status code: {response.status_code}",
+                )
+            else:
+                logger.debug(
+                    "External link %s successful in file %s, response code: %i",
+                    link,
+                    file,
+                    response.status_code,
+                )
+                return None
+        except requests.exceptions.ConnectionError as err:
+            logger.info(
+                "External link %s in file %s raised a connection error", link, file
+            )
+            return LinkReport(
+                "External link %s in file %s raised a connection error %i", err.errno
+            )
+
+    def _get_os_path(self, path: str) -> str:
+        """Gets an os-specific path from a path found in a markdown file"""
+        return path.replace("/", os.path.sep)
+
+    def _get_absolute_path(self, path: str) -> str:
+        return os.path.join(self._docs_root, self._get_os_path(path))
+
+    def _get_relative_path(self, file: str, path: str) -> str:
+        # link should be relative to the location of the current file
+        directory = os.path.dirname(file)
+        return os.path.join(directory, self._get_os_path(path))
+
+    def _get_docroot_path(self, path: str) -> str:
+        return os.path.join(self._docs_root, self._get_os_path(path))
+
+    def _check_absolute_link(self, link: str, file: str, path: str) -> LinkReport:
+        logger.debug("Checking absolute link %s in file %s", link, file)
```

Comment:
Same with these other type hints. I think they should all be `Optional[LinkReport]`?

Reply:
done!